src/library/base/R/connections.R

```diff
@@ -91,10 +91,17 @@ fifo <- function(description, open = "", blocking = FALSE,
 
 url <- function(description, open = "", blocking = TRUE,
                 encoding = getOption("encoding"),
-                method = getOption("url.method", "default"))
+                method = getOption("url.method", "default"),
+                headers = NULL)
 {
     method <- match.arg(method, c("default", "internal", "libcurl", "wininet"))
-    .Internal(url(description, open, blocking, encoding, method))
+    if (!is.null(headers)) {
+      if (length(names(headers)) != length(headers) || any(names(headers) == ""))
+        stop("'headers' must must have names")
+      headers <- paste0(names(headers), ": ", headers)
+      headers <- list(headers, paste(headers, "\r\n", collapse = ""))
```

Comment:
Do you mean to add a space at the end of each header line? e.g. using `paste()` instead of `paste0()`?

Also this structure is pretty weird, the headers with an without the trailing newlines? Why can't you pass one or the other and add the newlines internally (or vise-versa).

Reply:
Yeah, I mean that, there is a `\r\n` at the end.

Yeah, it is weird. The problem is that some download methods require a single collapsed string, others a list of strings. It is quite complicated to manage the memory if I want to transform one format to the other in C.

So the simplest is to just pass both.

Reply:
But you mean there to also be an extra space in between the text and the `\r\n`?

Reply:
You're currently adding `" \r\n"` though — that doesn't seem right.

Reply:
OH, yeah, you are right, that's not needed, `paste0` is better.

Reply:
Those spaces are actually ignored, but yeah, `paste0` is better.

Reply:
Btw. re passing the same info twice, I am open to better ways, if there is any.